File: concepts/04_collections_and_iteration/07_sorting_stability_and_custom_order/julia/test_02_key_failures_and_comparator_contracts.jl
# polyglot-family: collections_and_iteration
# polyglot-concept: sorting_stability_and_custom_order
# polyglot-related: languages/julia/language/06_arrays_indexing_iteration_and_broadcast/
# polyglot-related+: test_048_mapping_reduction_and_stable_sorting.jl
#
# 共同问题：key 提取失败怎样传播；自定义 comparator 必须满足什么契约。
# 对照观察：by/lt 的异常直接传播；lt 应定义 strict weak order，反向数值顺序可用 isless 参数交换。

using Test

@testset "排序不吞掉 key 错误" begin
    @test_throws FieldError sort([1, 2]; by = value -> getproperty(value, :missing))
    descending = sort([1, 3, 2]; lt = (left, right) -> isless(right, left))
    @test descending == [3, 2, 1]
    @test sort(["bbb", "a", "cc"]; by = length) == ["a", "cc", "bbb"]
end

File: concepts/04_collections_and_iteration/07_sorting_stability_and_custom_order/lua/test_02_key_failures_and_comparator_contracts.lua
-- Common question: how do key/comparator failures and invalid contracts surface?
-- Inputs: missing fields, comparator errors, mixed incomparable values, and side effects.
-- Observations: immediate propagation, partial mutation allowance, strict-order responsibility, and no key callback.
-- polyglot-family: collections_and_iteration
-- polyglot-concept: sorting_stability_and_custom_order
-- polyglot-related: languages/lua/standard_library/10_tables_math_and_random/test_074_sorting_and_custom_order.lua

local t = require("support.assertions")

local values = {{rank = 2}, {}, {rank = 1}}
t.raises(function()
    table.sort(values, function(left, right)
        assert(left.rank and right.rank, "missing rank")
        return left.rank < right.rank
    end)
end, "missing rank")

t.raises(function()
    table.sort({1, "2"})
end)

local comparisons = 0
table.sort({3, 2, 1}, function(left, right)
    comparisons = comparisons + 1
    return left < right
end)
t.truth(comparisons > 0)

-- Lua 只接受 comparator，没有独立 key function；比较器必须实现一致的严格顺序。
t.done()
